### ISSUE TITLE
Bidi streaming use cases for grpc-web

### ISSUE THREAD

**Issue by wenbozhu** [grpc/grpc-web#24]
Options, use cases to enable bidi streaming support with grpc-web

**Comment by DieMyst** [comment]
Hello, I want to write about my use case of bidi streaming.
We are in the process of writing decentralized database. 
So, all data is encrypted and client must follow the b-tree with encrypted keys step-by-step to decrypt each key and say in which direction should be next step.
We had good start in implementation this logic with grpc, but browser implementation without some wrappers (websocket, etc) of client is difficult now without bidi support.
Our goal is the ability to use a decentralized database via browser having only a private key available.

**Comment by johanbrandhorst** [comment]
Note that Imorobables grpc-web implementation has a beta bidi
implementation over websockets. https://github.com/improbable-eng/grpc-web.

**Comment by shanshanzhu** [comment]
@johanbrandhorst the improbable beta bidi version is really premature. It is not documented and has no example code. All it does is just exporting a transport function with websocket, but doesn't really wire it up with grpc. If I miss something (i.e. branch) there, could you help point me to the right place of code? Thanks!

 https://github.com/improbable-eng/grpc-web/blob/709b592779c35e17f581f9bf866f4cac4b0b93fb/ts/src/transports/Transport.ts#L64

**Comment by johanbrandhorst** [comment]
@shanshanzhu the websocket proxy implementation is good. It wraps the Go http.ResponseWriter and io.Reader interfaces with websocket read/writes and passes the request to the gRPC Go implementation, so the wire protocol is pure gRPC-over-http2-over-websockets. Take a look at the implementation, it's beautiful :heart_eyes: https://github.com/improbable-eng/grpc-web/blob/72eb701d6f320ca324b3347c7925a720b553eae5/go/grpcweb/wrapper.go#L133, https://github.com/improbable-eng/grpc-web/blob/709b592779c35e17f581f9bf866f4cac4b0b93fb/ts/src/transports/websocket.ts#L51.

Furthermore, as I already pointed out in https://github.com/improbable-eng/grpc-web/issues/205, you can look at the tests for an example of how to use it: https://github.com/improbable-eng/grpc-web/blob/master/test/ts/src/client.websocket.spec.ts#L77.

The GopherJS gRPC-Web bindings have been using the websocket proxy on https://grpcweb.jbrandhorst.com since the start of the year.

**Comment by wenbozhu** [comment]
Re: websockets, we would welcome all the gateways that provide inter-operability with gRPC end-points.

WS over HTTP/2 is being finalized and will be available in Chrome. We will evaluate if it makes sense to make WS a transport for grpc-web (but the bar to deploy this for Google's web apps will be high) and there will be details such as ping-pong, half-close etc to spec out ...

**Comment by nhooyr** [comment]
@johanbrandhorst does this mean there is a client side grpc-http2 implementation thats reading everything from the web sockets?

**Comment by nhooyr** [comment]
@wenbozhu My use case for bidirectional streaming is for a online game where the client is sending message based on what the user is doing and the server is sending messages to clients to broadcast all events. I could just do this over plain web sockets with protobuf but it'd definitely be nice to see support for this. Can we implement something similar to what @johanbrandhorst noted in improbable's grpc-web repo? It seems like it'd be really simple given its just a wrapper around the protocol and we could document that bidirectional streaming is done over web sockets. Later on, with WS over HTTP/2, we could enable support for that and then later for the streams API. This approach would ensure broad browser compatibility.

**Comment by johanbrandhorst** [comment]
I don't think we should implement this in this repo until we have request streaming in Fetch. There is a third party, non-spec implementation that users can use if they need bidi streaming before then.

**Comment by nhooyr** [comment]
@johanbrandhorst the issue with that is Fetch will only work on bleeding edge browsers. I think its a pragmatic idea to include WebSockets over HTTP/1.1 as a fallback.

**Comment by johanbrandhorst** [comment]
I disagree. This will by necessity be bleeding edge tech. The spec talks about eventually speaking the HTTP2 protocol, which is not supported by todays browsers. I don't think the aim of this project is to provide backwards compatibility to old browsers.

**Comment by wenbozhu** [comment]
grpc-web does intend to support important browser use cases, i.e. features, versions, environments etc ... and the current spec is implementation details but important.

I don't necessarily agree tunneling grpc/http2 over websockets is a good or sane idea, and you are going to run into issues such as flow-control, half-close etc, to name a few.

One of the problems with websockets is that connectivity is not guaranteed (e.g. to support gmail) .. which means you always need a fallback. We don't yet want this level of complexity for grpc-web.

I presented some chrome stats last year to a bunch of HTTP "vendors" 

https://docs.google.com/a/google.com/presentation/d/e/2PACX-1vTlBG1NyTKPpAAGTyChKRSPfleDP2UIyC1Jw9_0h7Z2DZL4qcZ6S3AIF0d_RNGjq5SFenyEQAP9HVrq/pub?start=false&loop=false&delayms=3000

===

This bug is intended to collect use cases (title updated :) ... e.g. the case @nhooyr mentioned, could the API be actually better (e.g. resilient to failures) if client uses unary calls to post and uses server-stream to receive messages?

**Comment by johanbrandhorst** [comment]
@wenbozhu 403 on that document Im afraid

**Comment by nhooyr** [comment]
> I don't necessarily agree tunneling grpc/http2 over websockets is a good or sane idea, and you are going to run into issues such as flow-control, half-close etc, to name a few.

Couldn't flow control be implemented at the grpc-web level? Likewise with half close?

> One of the problems with websockets is that connectivity is not guaranteed (e.g. to support gmail) .. which means you always need a fallback. We don't yet want this level of complexity for grpc-web.

In what way is the connectivity of WebSockets not guaranteed?

> e.g. the case @nhooyr mentioned, could the API be actually better (e.g. resilient to failures) if client uses unary calls to post and uses server-stream to receive messages?

The issue with that would be the constant authentication checks on every single message (of which there would be a lot) when doing an RPC. A bidirectional stream completely avoids this by only authenticating the connection at the start. I guess I could give each client a token before they start a game and then they pass that token on each request and we just directly compare that token instead of hashing for performance reasons on each RPC but still its more complicated than just a bidi stream.

Furthermore, while unary RPC+server stream avoids depending on the client side streaming connection remaining open, its not really more resilient because it'd be very easy to just reopen the bidi stream and we still have to worry about the server stream closing anyway. I don't think its more resilient in my case.

**Comment by nhooyr** [comment]
> Re: websocket connection failures, please check the doc on chrome stats. We don't really understand all the root causes, e.g. virus detection software etc.

If that happens, there should be no fallback. Just throw an error. Its unfortunate but welp. I think in the vast majority of cases, web sockets work fine.

> Re: wire semantics, yes, grpc-web specific concepts need be introduced than just tunneling grpc/http2 over websockets (as a TCP-like bytestream). This to me adds unnecessary complexities than doing a grpc-web specific mapping to websockets (as a transport).

Wouldn't grpc/http2 over web sockets take care of half close and flow control?

**Comment by wenbozhu** [comment]
grpc/http2 only passes the payload over websockets, not http/2 itself .. but if the latter is the case, I really don't know how to reason about this.

In any case, I'd agree that ws over http2 will make ws a viable transport for grpc-web (only).

**Comment by nhooyr** [comment]
> grpc/http2 only passes the payload over websockets, not http/2 itself .. but if the latter is the case, I really don't know how to reason about this.

Ah that sounds fine to me. I see what you mean now. We'd need a small wrapper protocol around websockets as well.

**Comment by ByteFlinger** [comment]
Just passing by to add a use case for client streaming.

I was looking to be able to support an API for file upload. From what I can gather, uploading a (maybe large) file would require client streaming.

I looked further into adding an HTTP endpoint to the gRPC server in my Go application but it seems that the http endpoint supported by the gRPC Go library requires an http/2 connection and TLS (which is not ideal if you are fronting your app with a load balancer doing TLS termination) so I might as well write a REST endpoin instead of gRPC.

I am curious as to why this specific use case was not mentioned, maybe I missed another way of doing that?

**Comment by johanbrandhorst** [comment]
AFAIK file uploads in the browser are most appropriately handled with multi-part upload requests, a well established web standard. Furthermore, I don't think there's a clear mapping between it and gRPC methods, unfortunately. An interesting thought would be if the gRPC-Web proxy could automatically translate such requests to a client streaming request of some well defined format.

**Comment by wenbozhu** [comment]
+1 on what @johanbrandhorst  said ....  We have no plan (or mechanism) to support random media tunneling as part of grpc-web, but this could be a "grpc gateway" feature for developers who only want to deal with grpc end-points.

**Comment by leonfancy** [comment]
I'm working on a speech regonition project, browser records and streams audio binary data to gRPC server and server streams the intermediate result to client.
```
rpc recognize(stream Audio) returns (stream Text) {}
```

In this case, bidi streaming is essential.

**Comment by imans777** [comment]
@leonfancy Same here

**Comment by micfan** [comment]
This feature finally will be implemented, then gRPC can own half of the share in the Browser side, just like what Node.js to JavaScript. No HTTP1 proxy, no WebSocket bridge, seamlessly combaine front and backend, no RESTful any more. Use cases are not important, what you want, what we have

**Comment by PavelSosin** [comment]
For those who adopt already released LSP and DAP protocols non-bidirectional gRPC without both client  and server push features doesn't make any sense. Both protocols are full duplex protocols built on the top of WebSocket and JSONRPC. Even. If server is implemented as Web/NodeWorkers with near zero communication protocol  latency transfer time for very big objects or long stream shifts end-user latency above allowed.  It is not a rare case when the final object is created by parallel threads or reactive streams. Wise developer can always spread work among workers keeping single communication channel. But the result serialization and transfer time is critical! Compression of transfer buffer using plane JSZIP gave me more than 10 folds improvement in communication time between browser/nodejs main thread and worker. I expect the same and, even higher, improvement from gRpc using compressed protobuf. Unfortunately buffer compression object is available in Nodejs 10+ but not in Chrome and Edge browsers built on the same V8 engines.

**Comment by emmveqz** [comment]
@micfan  when?  where is it (going to be)?  I need it!!

> This feature finally will be implemented

**Comment by micfan** [comment]
@emmveqz These text is cited from some docs. I have turn to the raw gRPC. If there no so many requests for this feature, can push it off. Or, ProtoBuffer over WebSocket is OK.

**Comment by wenbozhu** [comment]
@leonfancy Yes, speech recognition  is a canonical use case. 

We will publish a detailed plan soon (by Nov.), i.e. what choices we have, their limitations, timelines, as well as considerations of future Web transports. E.g. it could be the case that we decide to ship a basic websocket based solution to unblock the development environment first if this is what everyone wants.

**Comment by hochhaus** [comment]
@wenbozhu It isn't the end of nov yet, but I am very excited about the bidi grpc possibility. Any update here?

**Comment by wenbozhu** [comment]
Sorry about the delay. Will get something published before the year end.

**Comment by wenbozhu** [comment]
Not the exact plan yet but please take a look: [Road-map feedback](https://docs.google.com/forms/d/1sHKGMVlyDAzzVd34kOXfTvMvIcN1X2zRqVf56uFYcPs/edit)

**Comment by coryvirok** [comment]
Bump. :)

Am I correct in assuming that this functionality is blocked by Fetch full/half-duplex support? I'd be in favor of getting a working solution now and swapping it out for the more correct solution later. Considering this issue has been open since 2016, I'd say bleeding edge browser support is also OK for now since it will unblock many of us. 

Thanks for all of the hard work on this.

**Comment by wenbozhu** [comment]
We are in active discussion with chrome/blink on the future streaming Web APIs, e.g. request-streaming via fetch, web-transport/quic. At this point, I feel any effort on creating a formal websocket based solution is not a good investment. I will post a "road-map in progress" shortly. Thanks.

**Comment by wenbozhu** [comment]
A quick update:

1. fetch request-streaming: we will help run some experiments with Chrome to address certain compatibility concerns. We hope that most browsers will add fetch request-streaming support soon.
1. web-transport: the IETF and Web API standards may take some time to finalize but Chrome will have early support (over quic).

With the development of the above new Web APIs, our current plan is not to invest on any websocket-based solution to minimize the complexity of grpc-web.

Will keep you updated on the progresses we are making.

**Comment by glerchundi** [comment]
@wenbozhu thanks for the update. It helps keeping hope on this issue and I second your strategy on going through QUIC instead  of WebSockets.

**Comment by coryvirok** [comment]
I realize these types of features can take a while to implement, but it would be helpful to get a ballpark timeline for this. Is this likely going to take another month/6mo/year/more to implement? Is there anything the community can help out with to move this along?

**Comment by ydesgagn** [comment]
I would be willing to allocate 1-2 engineers on this to create a PR. Any guidelines we should follow so the PR can be accepted smoothly?

**Comment by jsmouret** [comment]
what about a webrtc data channel instead of websocket?

**Comment by wenbozhu** [comment]
Re: community help, I think we'd be happy to review any spec proposals to support Web APIs such as websockets, and to help convince the Envoy community to accept a module to proxy bidi Web traffic to grpc services. However, we will not be able to support them as part of the grpc-web JS client, i.e. if no Google service will be using such a solution.

Re: timelines, the Chrome origin trials are happening in Q3. Maybe we will have something by Q4.

Re: webrtc data channels, I am not familiar with the use of webrtc for client-server communication, or if the JS API is available for regular web apps. Do you have any references on the use of webrtc datachannel for bidi client-server communication?

Thanks!

**Comment by blockspacer** [comment]
@wenbozhu 

nice blog post about webrtc for client-server communication
http://blog.brkho.com/2017/03/15/dive-into-client-server-web-games-webrtc/

code based on blog post from blog.brkho.com, but with updated C++ client version + boost.beast (websockets) for connection establishing
https://github.com/lineCode/webrtc-datachannels

conan package with docker container, may be used for webrtc builds (tested on ubuntu 18.04)
https://github.com/blockspacer/conan-webrtc

**Comment by jsmouret** [comment]
I did a simple proof of concept at https://github.com/jsmouret/grpc-over-webrtc
More documentation to come but it is a start :)

**Comment by wenbozhu** [comment]
Thanks for the information on webrtc.  

@jsmouret  feel free to file a separate issue on webrtc, e.g. maybe we could build something under https://github.com/grpc-ecosystem.

**Comment by wenbozhu** [comment]
We have post a [road-map doc](https://github.com/grpc/grpc-web/blob/master/doc/streaming-roadmap.md) on our 2020 plan to improve the streaming support in grpc-web.  Feel free to post any questions here or file a separate issue to discuss. 

@ydesgagn  for websockets, you may want to contribute something under https://github.com/grpc-ecosystem first. I am happy to review any spec on grpc/websockets.

**Comment by kshcherban** [comment]
Use cases:

- stream request (microphone input, audio, video) from client  in browser (javascript) and receive back results (audio, video, text) from server (grpc-go), it can be any type of video, voice chat, speech to text or text to speech, etc.

**Comment by chwzr** [comment]
the roadmap looks very good! For my understanding: we are waiting for https://www.chromestatus.com/feature/5274139738767360 this to be available in major browser, to see the bidi-streaming magic happening here? 

I really like the idea of not using any websockets in the future, just pure grpc / http.
Please let us know if there are any updates on this :)

**Comment by chris-kruining** [comment]
upload streaming has landed in chrome 95. (https://web.dev/fetch-upload-streaming/). Any change this can be picked up, perhaps even soon?

**Comment by loeffel-io** [comment]
Webtransport support also shipped

**Comment by dunkymole** [comment]
An update on this this would be great.

**Comment by chris-kruining** [comment]
> An update on this this would be great.

not as much an update as a note, I implemented a gRPC lib that uses both upload and download streaming (in my lib that is, somewhere in the stack it could still buffer, so no promises on actual duplexing, but I should be supporting it), you could take a look at that and see it if is usable for what you are doing.

A couple of notes / warnings:
This is a new project, just 2 weeks old, so expect bugs (Issues and PR's are very much welcome).
Whilst I am using this in the development for my app (and is working with what I use it for), I am the sole user as far as I know, ergo: no community.
This lib implements gRPC, not gRPC-web, beware of that!

So, use on your own risk!

gRPC lib
https://github.com/FYN-Software/grpc

protoc plugin to generate .ts files for your .proto files that is compatible with this lib
https://github.com/FYN-Software/protoc-plugin-ts

docker container running the plugin and protoc
https://hub.docker.com/r/chriskruining/protoc-all

**Comment by edaniels** [comment]
> 

There's https://github.com/viamrobotics/goutils/tree/main/rpc/examples/echo which has an example providing an RPC server over grpc, grpc-web, grpc-gateway and grpc-over-webrtc.

**Comment by loeffel-io** [comment]
any update?

**Comment by sampajano** [comment]
Thanks for the interest! No work is currently planned right now.. But will update if they are. Thanks! :)

**Comment by dyaskur** [comment]
So many people interested for this feature. Why is there no plan for it? May I know the reason?

**Comment by sampajano** [comment]
@dyaskur Hi thanks for the interest. I'll provide an update in the roadmap soon thanks!

**Comment by vladchuk** [comment]
Full support (including bi-di streaming) of gRPC in Flutter seems like a glaring omission indeed. And still no update on the road map - will there be such support at all?

**Comment by sampajano** [comment]
An update to the roadmap is provided in https://github.com/grpc/grpc-web/pull/1326. Thanks all for the patience and support! 😃

@vladchuk FYI :)

**Comment by bitnom** [comment]
need it before I start using in server apps

**Comment by sampajano** [comment]
@pbower Thanks for checking! The bidirectional streaming support based on WebTransport is not currently being worked on yet but it's still planned as written. :)

We're making some changes soon to grpc-web including Typescript-ification so we'll probably get to that some time afterwards :)

**Comment by sampajano** [comment]
@Lopamoko Hi! Thanks for the interest and for checking! I cannot honestly give you a concrete date when this work will be started, since it's only planned but not really scheduled for this quarter or the next. Right now, i'm busy working on Typescript migration (first internal and external)which i'm hoping to finish later this year, and then I plan to improve on the set of fetch/stream performance / runtime related issues, before I have time to evaluate the bidi solution here. My optimistic hope is that I can start early next year. Hope this satisfies your question. :)

**Comment by popoolasubomi** [comment]
Hi guys,  are there still any efforts to support bi-directional streaming? My use case involves a search bar with search suggestions based on text updates. I need the server to return a list of new query suggestions as the user changes text on the search bar

**Comment by sampajano** [comment]
@popoolasubomi Hi! Thanks for checking! My [last comment](https://github.com/grpc/grpc-web/issues/24#issuecomment-2153734827) is still mostly up-to-date status wrt bidi support. We weren't able to start any work on it yet because Typescript migration is still underway. I will update this thread when we begin to work on it. Thanks for your patience!

**Comment by popoolasubomi** [comment]
Do you have any recommendations we could use in the spare time? Till bidirectional streaming is complete? I heard about web transport with grpcs

**Comment by sampajano** [comment]
@popoolasubomi No sorry nothing from us (on web) unfortunately :) WebTransport is on our roadmap but it's not been implemented yet (it needs server (Envoy) support as well).

**Comment by oh-tarnished** [comment]
Hello all,

All our systems were written with full gRPC support. Recently, we started migrating our application to Svelte from Flutter. Midway through the process, we realized that the browser doesn't support all gRPC lifecycles. Here’s the workaround we implemented using tRPC.

## Scenario:
You have a gRPC server written in any language (most of our workarounds and supported gateways are in `Go`). Then, there's a Node-based server (tRPC) which contains the logic for the gRPC client (Node) and connects to the tRPC client on the browser (Svelte).

> **Note:** This isn't the best approach for scalability. Since our application is intended to run locally, it works for us. However, I’m open to working on a more scalable solution if needed.

Please feel free to try it out and see how it goes!

## Demo:
[GitHub Repository](https://github.com/oh-tarnished/grpc-trpc-svelte)

**Comment by sourishkrout** [comment]
[Runme](https://runme.dev)'s notebook terminals use gRPC bi-directional streaming to sync stdio between processes. We'd love to switch to `grpc-web` and move off L4 TCP as a transport. While it looks like [Safari remains a holdout](https://developer.mozilla.org/en-US/docs/Web/API/WebTransport), all other browsers have `WebTransport` available.

Would it make sense to update the [Bidi Roadmap](https://github.com/grpc/grpc-web/blob/master/doc/roadmap.md) to make WebTransport available as experimental?

In my case, Runme runs containerized inside Electron (Chromium/Node.js runtime) and wouldn't require Safari support. Apple's a bit of an immovable object that won't have to block incremental progress here. No?

**Comment by Jeff92316046** [comment]
I'm very interested in the implementation of WebTransport in gRPC. Are there any updates on this?

**Comment by emmveqz** [comment]
Hi everyone,
I just started this [grpc-web-native](https://github.com/emmveqz/grpc-web-native)
It takes care of:
 - Streaming (request/response) (using `ReadableStream`)
 - Binary support (using `Uint8Array`)
 - No need for proxy (totally native)
 - Using HTTP2 with the browser native `fetch` (no websockets or `WebTransport` needed)

With the help of the already great npm pkg [protobuf.js](https://github.com/protobufjs/protobuf.js)

Take a look if you are interested. Suggestions and thoughts, or collaboration is more than welcome.

**Comment by luca992** [comment]
Looks like WebTransport is testable in the latest safari
https://github.com/mdn/browser-compat-data/issues/26151

**Comment by Lopamoko** [comment]
Hi! 👋
I just wanted to kindly check in: how are things going on this issue for @sampajano?
Has there been any clarity on when he might be able to take a look at it?

There’s a lot of interest in this feature from the community — many people are looking forward to it, and it could provide a significant boost to messaging-related services overall.

Thanks so much for all your work!

**Comment by luca992** [comment]
What tasks need to be completed to make this possible through prs? Are there issues on it?

**Comment by sampajano** [comment]
Hi :)

Thank you all for your enthusiasm and engagement around this feature over the past few years — we truly appreciate it.

However, as mentioned in https://github.com/grpc/grpc-web/pull/1484, we've made the decision not to pursue new major features in gRPC-Web, including bidi-streaming.

We sincerely apologize for this change and hope for your understanding.

Thank you all again :)

**Comment by edaniels** [comment]
Thanks for the update! I think it'd be prudent of Google to put a deprecation or unmaintained notice on this project at this point if it can't ever fully support gRPC. Especially since there are alternatives.

That being said for others reading this, implementing bidi via WebTransport, WebSockets, and WebRTC DataChannels is fairly easy to implement (I've done it with webrtc).
